### ISSUE TITLE
Undetermined or unsupported Byte Order on OpenBSD

### ISSUE THREAD

**Issue by feinerer** [mrash/fwknop#257]
Hi,

when compiling fwknop 2.6.9 on OpenBSD (6.2-beta; amd64) the byte order is not correctly detected.

OpenBSD ships with `/usr/include/endian.h` and `/usr/include/sys/endian.h` which are both detected via configure:
```
checking endian.h usability... yes
checking endian.h presence... yes
checking for endian.h... yes
...
checking sys/endian.h usability... yes
checking sys/endian.h presence... yes
checking for sys/endian.h... yes
```

Consequently `lib/fko_common.h` takes the path for `HAVE_ENDIAN_H` first:
```
#if HAVE_ENDIAN_H       /* Should cover most Linux systems */
  #include <endian.h>
  #define BYTEORDER __BYTE_ORDER
#elif
  ...
```
and defines `BYTEORDER` as `__BYTE_ORDER`. However, OpenBSD's [`endian.h`](https://github.com/openbsd/src/blob/master/sys/sys/endian.h) only defines `_BYTE_ORDER` (note the single leading underscore).

A potential (but ugly?) patch to address this issue was posted at https://marc.info/?l=openbsd-ports&m=150385082508265&w=2: [`patch-lib_fko_common_h`](https://marc.info/?l=openbsd-ports&m=150385082508265&q=p3).

Ideally you could include this patch (or variant of it) in the official sources. Thank you!

**Comment by mrash** [comment]
Thanks, that patch looks fine to me. I'll do a bit of testing over the next day or so and apply it.

**Comment by feinerer** [comment]
Jérémie Courrèges-Anglas (the author of the original patch) sent me a mail with additional information which I would like to share (with his permission):

>Various systems out there now implement a reasonable API through `/usr/include/endian.h`: they define `BYTE_ORDER`, `LITTLE_ENDIAN` and `BIG_ENDIAN`.
>This has been discussed for inclusion in the next POSIX standard release: https://www.opengroup.org/austin/docs/austin_514.txt
> Thus, upstream could use the following strategy:
>```
>#ifdef HAVE_ENDIAN_H /* POSIX proposal, should cover most modern systems */
>   #include <endian.h>
>   #ifndef BYTE_ORDER
>     #ifdef _BYTE_ORDER
>       #define BYTE_ORDER _BYTE_ORDER
>     #elif defined(__BYTE_ORDER)
>       #define BYTE_ORDER __BYTE_ORDER
>     #endif
>   #endif
>#elif HAVE_SYS_ENDIAN_H
>  ...
>#endif
>```
>Another solution would be for upstream to rely on `AC_C_BIGENDIAN`, after all they're already using it in their `configure.ac`.  The problem is that some old autotools releases are broken in this regard: https://lists.gnu.org/archive/html/autoconf/2008-08/msg00030.html

**Comment by mrash** [comment]
Fixed in 10d31064. Many thanks to Ingo Feinerer and Jérémie Courrèges-Anglas.
